extension/pytree/pybindings.cpp

```diff
@@ -145,7 +145,8 @@ class PyTree {
           } else if (py::isinstance<py::int_>(key)) {
             s.key(i) = py::cast<int32_t>(key);
           } else {
-            pytree_assert(false);
+            throw std::runtime_err(
+                "invalid key in pytree dict; must be int or string");
```

Comment:
It'd be helpful to print the key type here, though the old code didn't do it either so not a blocker. Similar for other new `throw` messages in this PR.